cli/loader/src/lib.rs

```diff
@@ -220,6 +223,24 @@ impl Loader {
                     .and_then(|extension| {
                         self.language_configuration_ids_by_file_type.get(extension)
                     })
+            })
+            .or_else(|| {
+                let Ok(file) = fs::File::open(path) else {
+                    return None;
+                };
+                let reader = BufReader::new(file);
+                let Some(Ok(first_line)) = std::io::BufRead::lines(reader).next() else {
+                    return None;
+                };
+
+                self.language_configuration_ids_by_first_line_regex
+                    .iter()
+                    .find(|(regex, _)| {
+                        Regex::from_str(regex)
+                            .and_then(|r| Ok(r.is_match(&first_line)))
+                            .unwrap_or(false)
+                    })
+                    .map(|(_, ids)| ids)
```

Comment:
I think all this implementation should be improved by using a `RegexSet`: https://docs.rs/regex/1.9.3/regex/struct.RegexSet.html

Reply:
Does regexset implement Hash? That's the main issue

Reply:
It's an additional note about implementation efficiency, it can be ignored cause requires more redesign, I can do it by myself later.